apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Logging/GoogleLoggerScopeTest.cs

```diff
@@ -0,0 +1,101 @@
+            Assert.Null(GoogleLoggerScope.Current);
+            using (new GoogleLoggerScope("grandparent"))
+            {
+                Assert.Equal("grandparent => ", GoogleLoggerScope.Current.ToString());
+                using (new GoogleLoggerScope("parent"))
+                {
+                    Assert.Equal("grandparent => parent => ", GoogleLoggerScope.Current.ToString());
+                    using (new GoogleLoggerScope("child"))
+                    {
+                        Assert.Equal("grandparent => parent => child => ", GoogleLoggerScope.Current.ToString());
+                    }
+                }
+                Assert.Equal("grandparent => ", GoogleLoggerScope.Current.ToString());
+                using (new GoogleLoggerScope("other_parent"))
+                {
+                    Assert.Equal("grandparent => other_parent => ", GoogleLoggerScope.Current.ToString());
+                }
+                Assert.Equal("grandparent => ", GoogleLoggerScope.Current.ToString());
+            }
+            Assert.Null(GoogleLoggerScope.Current);
+        }
+
+        [Fact]
+        public async Task Multiple_Tasks()
+        {
+            Func<string, string, string, Task> func = async (grandparent, parent, child) =>
+            {
+                Assert.Null(GoogleLoggerScope.Current);
+                using (new GoogleLoggerScope(grandparent))
+                {
+                    Assert.Equal($"{ grandparent} => ", GoogleLoggerScope.Current.ToString());
+                    await Task.Yield();
+                    using (new GoogleLoggerScope(parent))
+                    {
+                        Assert.Equal($"{grandparent} => {parent} => ", GoogleLoggerScope.Current.ToString());
+                        await Task.Yield();
+                        using (new GoogleLoggerScope(child))
+                        {
+                            Assert.Equal($"{grandparent} => {parent} => {child} => ", GoogleLoggerScope.Current.ToString());
+                            await Task.Yield();
+                        }
+                        await Task.Yield();
+                        Assert.Equal($"{grandparent} => {parent} => ", GoogleLoggerScope.Current.ToString());
+                    }
+                    await Task.Yield();
+                    Assert.Equal($"{ grandparent} => ", GoogleLoggerScope.Current.ToString());
+                }
+                Assert.Null(GoogleLoggerScope.Current);
+            };
+
+            await Task.WhenAll(
+                    Task.Run(() => func("grandparent-one", "parent-one", "child-one")),
+                    Task.Run(() => func("grandparent-two", "parent-two", "child-two")),
+                    Task.Run(() => func("grandparent-three", "parent-three", "child-three")),
+                    Task.Run(() => func("grandparent-four", "parent-four", "child-four")),
+                    Task.Run(() => func("grandparent-five", "parent-five", "child-five")));
+        }
+    }
```

Comment:
When you set an `AsyncLocal` and then enter an async method or spawn a new task, the code within that other method gets a copy of the value basically, but if they write to it, only they and calls they make will see that new value. Once execution returns to the thing which called the method or spawned the thread, the old value will still be there, because it is in a different async context. Basically, you want to try things like this:
```csharp
using (new GoogleLoggerScope("original"))
{
    await Foo();
}

public async Task Foo()
{
     await Something();
     using (new GoogleLoggerScope("foo"))
     {
          DoSomething();
     }
}
```

And make sure things work as expected as far as whether or not the foo scope should have a parent of the original scope.